s3datarepository/iam.go

```diff
@@ -241,132 +422,292 @@ func (s *S3Repository) GrantAccess(ctx context.Context, id string, derivative bo
 		}()
 	})
 
-	log.Debugf("created role %s", roleName)
+	var instanceRoleAssociated bool
 
-	// create instance profile
-	instanceProfileOutput, err := s.IAM.CreateInstanceProfileWithContext(ctx, &iam.CreateInstanceProfileInput{
-		InstanceProfileName: aws.String(roleName),
-		Path:                aws.String(s.IAMPathPrefix),
-	})
-	if err != nil {
-		return nil, ErrCode("failed to create instance profile "+roleName, err)
-	}
+	// if the instance already has some other instance profile, we copy all policies
+	// from the existing profile into the new one and then disassociate the old profile
+	if instanceInfo.IamInstanceProfile != nil {
+		// we only have the instance profile arn, so let's extract the name
+		ipArns := strings.Split(aws.StringValue(instanceInfo.IamInstanceProfile.Arn), "/")
+		currentInstanceProfileName := ipArns[len(ipArns)-1]
+		if currentInstanceProfileName == roleName {
+			instanceRoleAssociated = true
+		}
 
-	// append instance profile delete to rollback tasks
-	rollBackTasks = append(rollBackTasks, func() error {
-		return func() error {
-			if _, err := s.IAM.DeleteInstanceProfileWithContext(ctx, &iam.DeleteInstanceProfileInput{InstanceProfileName: aws.String(roleName)}); err != nil {
-				return err
+		if !instanceRoleAssociated {
+			log.Infof("instance %s already has instance profile %s, will try to migrate existing policies", instanceID, currentInstanceProfileName)
+
+			// find out what role(s) correspond to this instance profile and what policies are attached to them
+			var ipOut *iam.GetInstanceProfileOutput
+			ipOut, err = s.IAM.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{
+				InstanceProfileName: aws.String(currentInstanceProfileName),
+			})
+			if err != nil {
+				return nil, ErrCode("failed to get information about current instance profile "+currentInstanceProfileName, err)
 			}
-			return nil
-		}()
-	})
 
-	// add role to instance profile
-	_, err = s.IAM.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
-		InstanceProfileName: aws.String(roleName),
-		RoleName:            aws.String(roleName),
-	})
-	if err != nil {
-		return nil, ErrCode("failed to add role to instance profile "+roleName, err)
+			// TODO: we are _not_ considering role inline policies at this point, should we?
+			var currentPoliciesArn []string
+			for _, r := range ipOut.InstanceProfile.Roles {
+				log.Debugf("listing attached policies for role %s", aws.StringValue(r.RoleName))
+
+				var attachedRolePoliciesOut *iam.ListAttachedRolePoliciesOutput
+				attachedRolePoliciesOut, err = s.IAM.ListAttachedRolePoliciesWithContext(ctx, &iam.ListAttachedRolePoliciesInput{
```

Comment:
these should take into account the truncated response

Reply:
AWS says you can have 10 policies attached to a role (https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-limits.html#reference_iam-limits-entities).
Do we still need to worry about this getting truncated if MaxItems is 100?

Reply:
strange... why do they even have the paging?  i guess not... 😆